Pranjal Sharma,Agata Kaczmarek, Jan Smoleń, Mateusz Stączek

**Heidelberg University**

**Data Science  Group**
    
Prof. Dr. Michael Gertz  

Ashish Chouhan, Satya Almasian, John Ziegler, Jayson Salazar, Nicolas Reuter
    
December 4, 2023
    
Natural Language Processing with Transformers

Winter Semster 2023/2024     
***

# **Assignment 3: “Transformers”**
**Due**: Monday, January 8, 2024, 2pm, via [Moodle](https://moodle.uni-heidelberg.de/course/view.php?id=19251)



### **Submission Guidelines**

- Solutions need to be uploaded as a **single** Jupyter notebook. You will find several pre-filled code segments in the notebook, your task is to fill in the missing cells.
- For the written solution, use LaTeX in markdown inside the same notebook. Do **not** hand in a separate file for it.
- Download the .zip file containing the dataset but do **not** upload it with your solution.
- It is sufficient if one person per group uploads the solution to Moodle, but make sure that the full names of all team members are given in the notebook.

***

## **Task 1: Diving into Attention** (3 + 4 + 4 + 1 = 12 points)

In this task, you work with self-attention equations and find out why multi-head attention is preferable to single-head attention.

Recall the equation of attention on slide 5-9 to compute self-attention on a series of input tokens. We simplify the formula by focusing on a single query vector $q \in R^d$, value vectors ($\{ v_1,v_2,...,v_i \},v_i \in R^d$), and key vectors ($\{ k_1,k_2,...,k_i \},k_i \in R^d$). We then have

$$
a_i=\frac{exp(q^Tk_i)}{\Sigma^n_{j=1}exp(q^Tk_j)}
$$

$$
 o= \Sigma^n_{i=1} a_i v_i
$$

with $a_i$ being the attention weight for query $q$ with respect to key $k_i$. Then the output $o$ is the new representation for the query token as a weighted average of value vectors with weights $a=\{ a_1,a_2,...,a_i \},a_i \in R^d$.
Answer the following questions with the help of the equations and the intuition behind attention that you learned in the class:



### Subtask 1: Copying  

1.   Explain why $a$ can be interpreted as a categorical distribution.
2.   This distribution is typically diffuse, where the mass is spread out between different values of $a_i$. Describe a scenario in which the categorical distribution puts all the weight on a single element, e.g., $a_j \gg \Sigma_{j\neq i}a_i$. What are the conditions on key and/or query for this to happen?
3. In this case of a single large $a$, what would the output $c$ look like and what it means intuitively?

In attention, it is easy to **copy** a value vector $v_i$ to the output $o$.





**Answer**




1. A categorical distribution is categorisation of probabilities assigned to a set of discrete categories and same is $a$. we would take a different approach here i.e. using behaviour of categorical distribution:
a. Discrete sample space: $a$ also have discrete sample space
b. Non-Ordered categories: probabilities might not be ordered in categorical distribution as when calculation $ai$ but $ai$ denotes the weight of vectors on basis of which we further focus on elements of the vector.
c. Sum of probabilities is 1: softmax function converts the dot product into probabilites where the sum is 1
d. No correlation: attention is distributed and does not depend on the other probabilities

2. The weight of a single element can be significantly higher in the case where model heavily focuses on a single element of input sequence. dot product 
$exp(q^T.k_j)$ is much larger than the dot products $exp(q^T.k_i) for i ≠ j$, the softmax function will magnify this difference, assigning a high probability to $a_j$. 

3. in this output O will be very close to $v_j$, suggesting that the query vector has a significant similarity with the key vector and high attention weight for the jth element. this intitutively means the element has unique features not available in other elements and rare usage. so when this word is used in input query it is highly distinguishable.

Ex. Spam filter




#### ${\color{red}{Comments\ 1.1}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 2: Averaging


Instead of focusing on just one value vector $v_j$, the Transformer model can incorporate information from multiple inputs. Consider the situation where we want to incorporate information from two value vectors $v_b$ and $v_c$ with keys $k_b$ and $k_c$. In machine learning one of the ways to combine this information is through averaging of vectors $o= \frac{1}{2}(v_b+v_c)$.  It might seem hard to extract information about the original vectors $v_b$ and $v_c$ from the resulting average. But under certain conditions, one can do so. In this subtask, we look at the following cases:

1. Suppose we know the following:


* $v_b$ lies in a subspace $B$ formed by the $m$ basis vectors $\{b_1, b_2, .. , b_m\}$, while $v_c$ lies in a subspace $C$ formed by the $p$ basis vectors $\{c_1, c_2, . . . , c_p\}$ (This means that any $v_b$ and $v_c$ can be expressed as a linear combination of their basis vectors).
*   All basis vectors have the norm 1 and are orthogonal to each other.
*   The two subspaces $B$ and $C$ are orthogonal, meaning $b_j^Tc_k=0$ for all $j$ and $k$.
* Given that $\{b_1, b_2, .. , b_m\}$ are both orthogonal and form a basis for $v_b$, we know that there exists some $d_1, ..., d_m$ such that $v_b=d_1 b_1+d_2 b_2+...+d_m b_m$. Use these $d\text{s}$ to solve this task.

Using the basis vectors $\{b_1, b_2, .. , b_m\}$, construct a matrix $M$ such that for arbitrary vectors $v_b$ and $v_c$ with the given conditions, we can use $M$ to extract $v_b$ from the sum of the vector $s = v_b + v_c$. In other words, construct an $M$ such that  $ Ms = v_b$ holds.


2. If we assume that
* all key vectors are orthogonal, i.e., $k_i^Tk_j=0$ for all $i \neq j$, and
* all key vectors have the norm 1.

Find an expression for the query vector $q$ such that $o \approx \frac{1}{2}(v_b+v_c)$. Justify your answer.

**Hint:** Use your finding in subtask 1 to solve part 2.

**Hint:** If the norm of a vector $x$ is 1, then $x^Tx=1$

**Hint:** Start with writing $v_b$ and $v_c$ as the linear combination of the bases.


**Answer**


1. 

Given that, $v_b={d_1 b_1+d_2 b_2+...+d_m b_m}$

Also, $s = v_b + v_c$

substituting $v_b$ in the equation

$s = d_1 b_1+d_2 b_2+...+d_m b_m + v_c$

$M$ is a basis vectors $\{b_1, b_2, .. , b_m\}$ and multiplying $M$ with $s$

M s = $\{b_1, b_2, .. , b_m\} s $

$={b_1(d_1 b_1+d_2 b_2+...+d_m b_m)+b_2(d_1 b_1+d_2 b_2+...+d_m b_m)+...+b_m(d_1 b_1+d_2 b_2+...+d_m b_m)}$

$= d_1 b_1+d_2 b_2+...+d_m b_m + v_c = v_b$ 

So, the matrix $M$ is constructed such that multiplying it by the sum vector $s$ allows us to extract the original vector ${𝑣_𝑏}$.


2. 

$M s = v_b$

substituting $s$

$M(v_b + v_c) = v_b$

$Mv_b + Mv_c = v_b$

$Mv_b = v_b - Mv_c$

using the orthogonal properti of basis vectors and find speciifc form of $M$:

$k_i^Tk_j = 0 \qquad \text{for } i \neq j$

$k_i^Tk_i = 1 \qquad \text{for all } i$

$c_i^Tc_j = 1 \qquad \text{for } i=j$

$c_i^Tc_k = 0 \qquad \text{for } i \neq k$

$k_i^Tc_k = 0 \qquad \text{for all } i,k$


$M$ = \begin{pmatrix}
2k_1^Tk_1 & k_1^Tk_2 & \dots & k_1^Tk_m \\
k_2^Tk_1 & 2k_2^Tk_2 & \dots & k_2^Tk_m \\
\vdots & \vdots & \ddots & \vdots \\
k_m^Tk_1 & k_m^Tk_2 & \dots & 2k_m^Tk_m
\end{pmatrix}


We have this, 

$o = \frac{1}{2}(v_b + v_c)$

Also,  $s = v_b + v_c$

$\therefore $ $o = \frac{1}{2}(s)$

Further we can say,
$\frac{1}{2}Ms = \frac{1}{2}M(v_b + v_c) = \frac{1}{2}v_b$



#### ${\color{red}{Comments\ 1.2}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 3: Drawbacks of Single-head Attention

You might have wondered why we need multi-heads for attention. In this subtask, we look at some of the drawbacks of having a single head attention. As shown in the previous subtask, it is possible for single head attention to focus equally on two values. The same can apply to any subset of values, which therefor can become problematic.

Consider a set of key vectors $\{ k_1,k_2,...,k_n \}$, randomly sampled from a normal distribution with a known mean value of $\mu_i \in R^d$ and unknown covariance $Σ_i, i \in \{1, \ldots, n\}$, where


*   $\mu_i\text{s}$ are all orthogonal $\mu_i^T\mu_j=0$ if $i \neq j$.
*   $\mu_i\text{s}$ all have unit norm $||\mu_i||=1$.

1. For a vanishingly small $\alpha$ (not to be confused with attention weights), the covariance matrices are  $Σ_i=\alpha I, \forall i  \in \{1,2,..,n\}$, design a query $q$ in terms of the $\mu_i$ such that as before, $o= \frac{1}{2}(v_b+v_c)$ and describe why it works.

2.  Large perturbations in key value might cause problems for single head attention.  Specifically, in some cases, one key vector $k_b$ may be larger or smaller in norm than the others, while still pointing in the same direction as $\mu_b$. As an example of such a case,
consider a covariance matrix for item $b$ for vanishingly small $\alpha$ as $Σ_b=\alpha I + \frac{1}{2}(\mu_b^T\mu_b)$. This causes $k_a$ to point to roughly the same direction as $\mu_b$ but with large differences in magnitude, while for other items. Further, let $Σ_i=\alpha I\  \forall_i i \neq b$. When you sample multiple keys from the distribution $\{ k_1,k_2,...,k_n \}$ and use the $q$ vector from the pervious part, what do you expect vector $o$ to look like? Explain why this shows the drawback of single-head attention.

**Hint:**
Think about how it differs from pervious part and how $o$'s variance would be affected by the change in $Σ_b$.

**Hint:** Considering that $\mu_b^T\mu_b=1$, think of what are the ranges $Σ_b$ can take and how does that effect a sampled $k_b$ value.

**Hint:** $\frac{exp(b)}{exp(b)+exp(c)}=\frac{exp(b)}{exp(b)+exp(c)}\frac{exp(-b)}{exp(-b)}= \frac{1}{1+exp(c-b)}$

**Answer:**


1. For a vanishing small $\alpha$ covariance matrices are approx. dialgonal therefore, query can be focused on two key vectors at the same time even if they are orthogonal.

$ q = \frac{1}{\sqrt{n}} \sum_{i=1}^n \mu_i $

The term $ q = \frac{1}{\sqrt{n}} $ ensures that the magnitude of the query vector remains normalized (unit norm) even when summing over all orthogonal vectors.

 $Σ_i=\alpha I $

The covariance matrices $Σ_i=\alpha I $ imply that the covariance between different components of the vectors is zero. Therefore, the mean vectors $\mu_i$ are effectively independent.

This query is a weighted sum of the key vectors, with each key vector being weighted equally. This means that the query will focus equally on all of the key vectors, regardless of their orthogonality.

2. $k_b$ varies between $(\alpha + 0.5)\mu_b$ and $(\alpha + 1.5)\mu_b$ because $\mu_b^T\mu_b=1$. Thus: $k_b \approx \lambda\mu_b$ where $\lambda \in [0.5, 1.5]$.

   $\alpha$ being vanishingly small, we can again assume that $k_i \approx \mu_i$ for all $i \neq b$. Since $\mu_i^T\mu_j=0$ for all $i \neq j$, dot product between $q$ and any key vector except $k_b$ and $k_c$ is $0$. For $k_b$ and $k_c$ the following happens:

    $k_b^Tq = \lambda\mu_b^T\beta(\mu_b + \mu_c)= \lambda \beta$

    $k_c^Tq = \mu_c^T\beta(\mu_b + \mu_c)= \beta$
    
    We can now compute the attention weights $a_b$ and $a_c$:

    $a_b = \frac{exp(\lambda\beta)}{n-2+exp(\lambda\beta)+exp(\beta)} \approx \frac{exp(\lambda\beta)}{exp(\lambda\beta)+exp(\beta)} \approx \frac{1}{1+exp(\beta(1-\lambda))}$

    $a_c = \frac{exp(\beta)}{n-2+exp(\lambda\beta)+exp(\beta)} \approx \frac{exp(\beta)}{exp(\lambda\beta)+exp(\beta)} \approx \frac{1}{1+exp(\beta(\lambda - 1))}$

    All other attention weights are close to zero, so we can ignore their influence on the output.

    When $\lambda = 0.5$, then $a_b \approx 0$ and $a_c \approx 1$. In this case, the output $o \approx v_c$.
    
    When $\lambda = 1.5$, then $a_b \approx 1$ and $a_c \approx 0$. Here the output is $o \approx v_b$.

     $o$ oscillates between $v_b$ and $v_c$.



 These effects can be seen as drawbacks of single-head attention.

a. The output vector o will be dominated by the contribution of the key k. This can make it difficult for the model to learn about the other key vectors.

b. The model may be more sensitive to noise in the key vectors. This is because small changes in the norm of k will have a large impact on the attention weights.



#### ${\color{red}{Comments\ 1.3}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 4: Model Size  
1. Imagine you have an input sequence of  $l$ tokens, how much memory is required and what time complexity do we have for a single self-attention layer? (give your answer in terms of $l$)
2. If you have $N$ layers of self-attention, how  would the memory requirements and the time complexity change? (give your answer in terms of $l$ and $N$)
3. If you have $l=10,000$ and $10$ layers, with the ability to perform $10M$ operations per second, how long would it take to compute the attention output?


**Answer**


1. The memory required for the input sequence of l tokens where the dimension of input vector is one - 
### l * l * 1 
because each attention head needs to be stored into a matrix of size l*l

Time complexity of this would be - 
### O(l*l)

2. The memory required for the input sequence of l tokens for N layers of self attention where the dimension of input vector is one -
### N* l* l* 1
which is still proportional to l*l

Time complexity of this would be -
### O(N* l* l)

3. l = 10000 and N=10 
Number of operations per second 10M = 10^7

Time required, O(l * l * N) = 10^4 * 10^4 * 10 /10^7 = 100seconds 

#### ${\color{red}{Comments\ 1.4}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

## **Task 2: Multiple Choice Question Answering** (4 + 3 + 5 + 2 = 14 points)

In this task, you will fine-tune a transformer model on a multiple-choice task, which is the task of selecting the most plausible inputs in a given selection. The dataset used here is [SWAG](https://www.aclweb.org/anthology/D18-1009/), which is available via the Hugging Face [hub](https://huggingface.co/datasets/swag). Check the link for an overview of the dataset. SWAG is a dataset about commonsense reasoning, where each example describes a situation and then proposes four options that could apply for it.
Let's start by installing the necessary packages.

In [ ]:
%pip install transformers
%pip install datasets
%pip install evaluate
%pip install accelerate -U
%pip install sentencepiece

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 507.1/507.1 kB 5.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 115.3/115.3 kB 6.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 11.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.1/84.1 kB 2.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 265.7/265.7 kB 5.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 9.2 MB/s eta 0:00:00


In this task, you will use a BERT model with a `MultipleChoice` head from the Hugging Face library and then create your custom model.   Recall from the class that the BERT model has an auxiliary next sentence prediction task, in which two sentences are given to BERT separated by a `[SEP]` token and a classifier head decides if the second sentence logically follows the first one. Hugging Face has
 a `*ForMultipleChoice` architecture that uses the representation of the `[CLS]` token and a linear layer to classify if one sentence follows the other. We first start with this default architecture and then build a more complicated one in a later subtask.

### Subtask 1: Loading and Processing the Data

We use the `dataset` library to download the SWAG dataset, which already contains train, validation, and test splits.

In [ ]:
import os
os.environ['CUDA_LAUNCH_BLOCKING'] = "1"
from datasets import load_dataset, load_metric
datasets = load_dataset("swag", "regular")
datasets

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:72: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/datasets/load.py:1429: FutureWarning: The repository for swag contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/swag
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(


Generating train split:   0%|          | 0/73546 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/20006 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/20005 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['video-id', 'fold-ind', 'startphrase', 'sent1', 'sent2', 'gold-source', 'ending0', 'ending1', 'ending2', 'ending3', 'label'],
        num_rows: 73546
    })
    validation: Dataset({
        features: ['video-id', 'fold-ind', 'startphrase', 'sent1', 'sent2', 'gold-source', 'ending0', 'ending1', 'ending2', 'ending3', 'label'],
        num_rows: 20006
    })
    test: Dataset({
        features: ['video-id', 'fold-ind', 'startphrase', 'sent1', 'sent2', 'gold-source', 'ending0', 'ending1', 'ending2', 'ending3', 'label'],
        num_rows: 20005
    })
})

Lets look at the first item to see how the data looks like:

In [ ]:
datasets["train"][0]

{'video-id': 'anetv_jkn6uvmqwh4',
 'fold-ind': '3416',
 'startphrase': 'Members of the procession walk down the street holding small horn brass instruments. A drum line',
 'sent1': 'Members of the procession walk down the street holding small horn brass instruments.',
 'sent2': 'A drum line',
 'gold-source': 'gold',
 'ending0': 'passes by walking down the street playing their instruments.',
 'ending1': 'has heard approaching them.',
 'ending2': "arrives and they're outside dancing and asleep.",
 'ending3': 'turns the lead singer watches the performance.',
 'label': 0}

**Question:**
Look at the dataset card on the Hugging Face hub and define what each of these fields means, with respect to the task:

*   `sent1`:
*   `sent2`:
*    `ending0`, `ending1`, `ending2` and `ending3`:
*   `label`:




**Answer**

*   `sent1`: is the first sentence that is given to the model
*   `sent2`: is the beginning of the second sentence, which should be filled
*    `ending0`, `ending1`, `ending2` and `ending3`: propositions of the endings of the second sentence, one of them is correct
*   `label`: information which of the above endings is correct


Write a function that displays the context and each of the four choices, following the format


```
Context:...
A-
B-
C-
D-
Ground truth: option ...
```

How you display the results is not important. You should be able to extract different parts of the data correctly and know what each field represents.

In [ ]:
def explain_example(example):
  ### your code ###
  print(f"Context: {example['startphrase']} \n A- {example['ending0']} \n B- {example['ending1']} \n C- {example['ending2']} \n D- {example['ending3']} \n Ground truth: option {example['label']}.")

  ### your code ###

In [ ]:
explain_example(datasets["train"][0])

Context: Members of the procession walk down the street holding small horn brass instruments. A drum line 
 A- passes by walking down the street playing their instruments. 
 B- has heard approaching them. 
 C- arrives and they're outside dancing and asleep. 
 D- turns the lead singer watches the performance. 
 Ground truth: option 0.


Before feeding the data into the model, we need to preprocess the text using `Tokenizer` to tokenize the inputs into tokens and put it in a format that the model expects. The tokenizer specific to the model we want to use for this task is `distilbert-base-uncased`. Complete the code below to load a fast tokenizer for this model. DistilBERT is similar to the BERT model, and we only use this particular architecture for faster training.


In [ ]:
from transformers import AutoTokenizer

###your code###
tokenizer = AutoTokenizer.from_pretrained('distilbert-base-uncased')
###your code###

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:72: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/483 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

In [ ]:
tokenizer("This is the first sentence!", "And this is the second one.")

{'input_ids': [101, 2023, 2003, 1996, 2034, 6251, 999, 102, 1998, 2023, 2003, 1996, 2117, 2028, 1012, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

Write a function that preprocesses the samples.
The tricky part is to put all the possible pairs of sentences in two big lists before passing them to the tokenizer.
Each **first** sentence has to be repeated 4 times to go with different ending options.
There should be a separator token between the first and second sentence, to follow the BERT input logic.
The final output is a list of 4 elements, one for each choice, where the input is transformed by the tokenizer.
For example, with a list of 2 training examples, the output includes 2 lists, where each contains 4 elements. Each of those elements is the converted input ID of the first sentence followed by the second sentence with different endings.
When calling the `tokenizer`, we use the argument `truncation=True`. This will ensure that an input longer than what the model selected can handle will be truncated to the maximum length accepted by the model.

**Hint:** Flatten the lists (all choices are flattened into a single list) before feeding them into the tokenizer and unflatten them once again for the final output.

In [ ]:
### your code ###
ending_names =['ending0', 'ending1', 'ending2', 'ending3']
### your code ###
def preprocess_function(examples):
  ### your code ###
    # repeat each first sentence four times
    first_sentences = [[examples['startphrase'][i] for _ in range(4) ]for i in range(len(examples['sent1']))]
    # second sentences possible are combination of header and ending
    question_headers =[[f"{inner} [SEP]" for inner in first_sent] for first_sent in first_sentences]
    helping = [[examples[ending][i] for ending in ending_names ]for i in range(len(examples['sent1']))]
    second_sentences=[[help + " " + end for help, end in zip(sublist1, sublist2)] for sublist1, sublist2 in zip(question_headers, helping)]

    # flatten everything
    flattened=[item for sublist in second_sentences for item in sublist]

    # tokenize
    inputs=tokenizer(flattened, truncation=True)
    # un-flatten
    num=len(ending_names)
    inputs_end={input_ids: [value[i:i+num] for i in range(0, len(value), num)] for input_ids, value in inputs.items()}

    return inputs_end
    ### your code ###

In [ ]:
examples = datasets["train"][:2]
print(examples)
features = preprocess_function(examples)
print(len(features["input_ids"]), len(features["input_ids"][0]), [len(x) for x in features["input_ids"][0]])# output should be 2 4 [30, 25, 30, 28]

{'video-id': ['anetv_jkn6uvmqwh4', 'anetv_jkn6uvmqwh4'], 'fold-ind': ['3416', '3417'], 'startphrase': ['Members of the procession walk down the street holding small horn brass instruments. A drum line', 'A drum line passes by walking down the street playing their instruments. Members of the procession'], 'sent1': ['Members of the procession walk down the street holding small horn brass instruments.', 'A drum line passes by walking down the street playing their instruments.'], 'sent2': ['A drum line', 'Members of the procession'], 'gold-source': ['gold', 'gen'], 'ending0': ['passes by walking down the street playing their instruments.', 'are playing ping pong and celebrating one left each in quick.'], 'ending1': ['has heard approaching them.', 'wait slowly towards the cadets.'], 'ending2': ["arrives and they're outside dancing and asleep.", 'continues to play as well along the crowd along with the band being interviewed.'], 'ending3': ['turns the lead singer watches the performance.', '

We can now apply our function to all the examples in the dataset. We use the `map` method to apply the function on all the elements of all the splits in the dataset (training, validation, and testing).
Note that we passed `batched=True` to leverage the fast tokenizer and use multi-threading to process the texts in batches concurrently.

In [ ]:
encoded_datasets = datasets.map(preprocess_function, batched=True)

Map:   0%|          | 0/73546 [00:00<?, ? examples/s]

Map:   0%|          | 0/20006 [00:00<?, ? examples/s]

Map:   0%|          | 0/20005 [00:00<?, ? examples/s]

Our dataset is still not converted to tensors and not padded. This is the job of the `data collator`. A data collator takes a list of examples and converts them to a batch.
There is no data collator in the Hugging Face default library that works on our specific problem. We thus need to write our own one. In this collator:

*  All the inputs/attention masks are flattened.
* A flattened list is passed to the `tokenizer.pad ` method to apply dynamic padding to pad inputs to the maximum length in the batch. Output will be the size of `(batch_size * 4) x seq_length`.
* Everything needs to be unflattened for the output of the data collator.
* `input_ids` and `labels` should be returned as tensors.
* The output is a dictionary called `batch` that contains features needed for training (`input_ids`, `attention_mask`, `label`).



In [ ]:
from dataclasses import dataclass
from transformers.tokenization_utils_base import PreTrainedTokenizerBase, PaddingStrategy
from typing import Optional, Union
import torch

@dataclass
class MultipleChoiceDataCollator:
    tokenizer: PreTrainedTokenizerBase
    padding: Union[bool, str, PaddingStrategy] = True
    max_length: Optional[int] = None
    pad_to_multiple_of: Optional[int] = None

    def __call__(self, features):
        accepted_keys = ["input_ids", "attention_mask", "label"]
        if len(features[0])>len(accepted_keys):
          features=[{k: v for k, v in i.items() if k in accepted_keys} for i in features]
      ### your code ###

        labels =[feature['label'] for feature in features]
        # flatten
        flattened_features = {key: [item for sublist in [feature[key] for feature in features]
                              for item in sublist] for key in accepted_keys[:-1]}

        # use the tokenizer and attributes from the class to pad the input
        batch = self.tokenizer.pad(
            flattened_features,
            padding=self.padding,
            max_length=self.max_length,
            pad_to_multiple_of = self.pad_to_multiple_of,
            return_tensors="pt"
        )

        # un-flatten
        num=len(features[0]["input_ids"])
        batch_1 = {key: value.clone().detach() for key, value in batch.items()}
        batch = {}
        for key, value in batch_1.items():
          grouped_values = [value[i:i+4] for i in range(0, len(value), 4)]
          batch[key] = torch.stack(grouped_values)
        batch["labels"]=torch.tensor(labels)

        ### your code ###
        return batch

In [ ]:
accepted_keys = ["input_ids", "attention_mask", "label"]
features = [{k: v for k, v in encoded_datasets["train"][i].items() if k in accepted_keys} for i in range(10)]
batch=MultipleChoiceDataCollator(tokenizer)(features)
print(batch["input_ids"].shape)
print(batch["attention_mask"].shape)
print(batch["labels"].shape)

You're using a DistilBertTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


torch.Size([10, 4, 40])
torch.Size([10, 4, 40])
torch.Size([10])


In [ ]:
labels =[feature['label'] for feature in features]
# flatten
flattened_features = {key: [item for sublist in [feature[key] for feature in features]
                      for item in sublist] for key in accepted_keys[:-1]}

In [ ]:
for i in range(4):
  print(batch["input_ids"][0][i])
  print(tokenizer.decode(batch["input_ids"][0][i]))

tensor([  101,  2372,  1997,  1996, 14385,  3328,  2091,  1996,  2395,  3173,
         2235,  7109,  8782,  5693,  1012,  1037,  6943,  2240,   102,  5235,
         2011,  3788,  2091,  1996,  2395,  2652,  2037,  5693,  1012,   102,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0])
[CLS] members of the procession walk down the street holding small horn brass instruments. a drum line [SEP] passes by walking down the street playing their instruments. [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]
tensor([  101,  2372,  1997,  1996, 14385,  3328,  2091,  1996,  2395,  3173,
         2235,  7109,  8782,  5693,  1012,  1037,  6943,  2240,   102,  2038,
         2657,  8455,  2068,  1012,   102,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,     0])
[CLS] members of the procession walk down the street holding small horn brass instruments. a drum line [SEP] has heard approaching them. [

#### ${\color{red}{Comments\ 2.1}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 2: Fine-tuning a Hugging Face Model

To fine-tune our model, we first need to download the correct architecture from Hugging Face. Import the correct class for this task and download the pre-trained checkpoint for the base class from `distilbert-base-uncased`. Note that the weights in the classification head are initialized at random.

In [ ]:
### your code ###
from transformers import AutoModelForMultipleChoice, DistilBertConfig
model_name = "distilbert-base-uncased"
config = DistilBertConfig.from_pretrained(model_name)
model_hf = AutoModelForMultipleChoice.from_pretrained(model_name, config=config)

### your code ###

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:72: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Some weights of DistilBertForMultipleChoice were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['pre_classifier.bias', 'classifier.weight', 'pre_classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Next, we need to define our `Trainer` and pass in the correct `TrainingArguments` (a class that contains all the attributes to customize the training). Define a `TrainingArguments` that


* creates an output directory `distilbert-base-uncased-swag` to save the checkpoints and logs.
*   evaluates the model on the validation set after the `300` steps.
* a checkpoint should be saved after each `600` step and no more than 2 checkpoints should be saved in total.
* the random seed for training is `77`.
* batch size for training and evaluation: `48` (if you are running out of memory, feel free to change this setting but indicate it as a comment in your notebook, on a T4 GPU from google colab this takes about `13.2GB` of `15.0GB`).
* train for `1800` steps with a learning rate of `5e-5`, and add weight decay of `0.01` to the optimizer.
* the trainer should remove the columns from the data that are not used by the model.
* The final checkpoint should be the checkpoint that had the best overall validation metric not necessarily the last checkpoint.

**Note:** Please use GPU for to train your model. If on colab, you can use T4 GPU for free.

In [ ]:
from transformers import TrainingArguments, Trainer
### your code ###
training_args = TrainingArguments(
    output_dir="./distilbert-base-uncased-swag",
    evaluation_strategy="steps",
    eval_steps=300,
    save_strategy="steps",
    save_steps=600,
    save_total_limit=2,
    seed=77,
    per_device_train_batch_size=48,
    per_device_eval_batch_size=48,
    max_steps=1800,
    learning_rate=5e-5,
    weight_decay=0.01,
    remove_unused_columns=True,
    load_best_model_at_end=True,
)

### your code ###


Before we initialize the `Trainer`, we create a function that tells the trainer how to compute the metrics from the predictions. Fill the `compute_metrics` function to compute the accuracy based on the `predictions`. This object contains the prediction of the model, as well as the ground truth labels.

**Hint 1:** Keep in mind that the output of this function should be a dictionary containing the metric name and value.

**Hint 2:** Consider the shape of the example input. This is similar to the logits produced by the model.

In [ ]:
import numpy as np
def compute_metrics(predictions):
   ### your code ###
   preds, label_ids = predictions
   predicted_labels=np.argmax(preds, axis=1)
   acc=sum(label_ids==predicted_labels)/len(label_ids)
   return_dict={"accuracy": acc}
   ### your code ###
   return return_dict

In [ ]:
preds=np.array([[0.9,0.2,0,0],
                [0.2,0.2,0.9,0.1],
                [0.2,0.9,0,0],
                [0.2,0.1,0.8,0],
                [0.9,0.1,0.8,0],
                [0.2,1,0.4,0],
                [0.2,1,0.4,0.9],
                [1,0.1,0.4,0.3],
                [0.1,0.1,0.9,0.3],
                [0.1,0.1,0.2,1]])
label_ids=np.array([0,3,1,2,0,1,3,0,2,3])
compute_metrics((preds,label_ids))

{'accuracy': 0.8}

Now it's time to pass everything to a `Trainer` object to start the training process. Initialize a `Trainer` object and pass all the necessary information, keep in mind that we also have the optional metric computation and that we tend to run an evaluation on the validation set during training. The training should take around 30 min on Google Colab T4 GPU.

In [ ]:
device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
device

device(type='cuda')

In [ ]:
### your code ###
trainer = Trainer(
    model = model_hf,
    args = training_args,
    train_dataset = encoded_datasets['train'],
    eval_dataset = encoded_datasets['validation'],
    tokenizer = tokenizer,
    data_collator = MultipleChoiceDataCollator(tokenizer),
    compute_metrics = compute_metrics
)
### your code ###

In [ ]:
trainer.train()# should take around 30 min on Google Colab T4 GPU

TrainOutput(global_step=1800, training_loss=0.9015062459309896, metrics={'train_runtime': 2490.0644, 'train_samples_per_second': 34.698, 'train_steps_per_second': 0.723, 'total_flos': 5678117920017504.0, 'train_loss': 0.9015062459309896, 'epoch': 1.17})

Save the model in `distilbert-base-uncased-swag/final_model`.

In [ ]:
### your code ###
trainer.save_model("./distilbert-base-uncased-swag/final_model")
### your code ###

Look at the saved files and answer the following questions (it is possible to answer these questions by writing some code, but we want you to explore the saved files):

**Question:**


1.   What is the vocabulary id for the `[CLS]` and `[MASK]` tokens?
2.   What is the dropout probability for the attention layer?

**Dropout:** With dropout, certain nodes are set to the value zero in a training run, i.e. removed from the network. Thus, they have no influence on the prediction and also in the backpropagation. Thus, a new, slightly modified network architecture is built in each run and the network learns to produce good predictions without certain inputs. Read more [here](https://databasecamp.de/en/ml/dropout-layer-en).



**Answer**


1. [CLS] is 102, [MASK] is 104
2. attention_dropout is 0.1


#### ${\color{red}{Comments\ 2.2}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 3: Fine-tune a Custom Model


In this case, we were lucky that Hugging Face had a pre-implemented architecture available for us to use. However, that is not always the case. Moreover, we might want to experiment beyond the default architectures to find a suitable one for a task. Therefore, it is important to learn to extend the Hugging Face models and train a custom model. The good news is that except for the model architecture the rest of the code can remain as it is.

Design a model for multiple choice model as follows:


1.   the config file for a feature extractor (must be a distilbert type) is  passed during initialization. The config file determines which model is used for feature extraction.
2.   From the `last_hidden_state` of the feature extractor, choose the `[CLS]` embedding (first one). This embedding is used as the compressed representation of first and second sentences. During pre-training it is used  for classifying whether these two sentences follow one another, making it a good candidate for our task.
3. `[CLS]` embedding is passed through a linear layer **that does not change the size of the embedding** and is passed through a tanh nonlinearity.
4. The output of tanh is passed through a dropout layer, where the dropout probability is the same as the dropout probability used for the `distilbert` model used as feature extractor.
5. The output of the previous stage is fed into another linear layer that shrinks the size of the embedding dimension to a quarter of the original size, e.g., if the embedding size is 12, the new embedding dimension is 3.
6. The output is followed by another dropout layer (you can use the one from stage 4).
7. Finally, a binary classifier is applied to determine the probability of sentence 1 being followed by sentence 2.
8. the cross-entropy loss is used to compute the loss.

**Hint:** Keep in mind that for a 4 choice system, you classify each of the four solutions independently. However, the final output should group the four logits together. For example, if input ids have the shape `[2, 4, 35]` (batch size=2, num choices=4, seq len=35), then the logits have the `[2, 4]` and labels have the dimension `[2, 1]`.



In [ ]:
from transformers import DistilBertModel,BertConfig,DistilBertConfig,PretrainedConfig,PreTrainedModel,DistilBertPreTrainedModel
from torch import nn

class CustomMultipleChoice(DistilBertPreTrainedModel):
    def __init__(self, config: PretrainedConfig):
        super().__init__(config)
        ###your code ###
        self.distilbert = DistilBertModel(config)
        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
        self.activation = nn.Tanh()
        self.dropout = nn.Dropout(config.seq_classif_dropout)
        self.dense2 = nn.Linear(config.hidden_size, config.hidden_size // 4)
        self.classifier = nn.Linear(config.hidden_size // 4, 1)
        ###your code ###


    def forward(
        self,
        input_ids: Optional[torch.Tensor] = None,
        attention_mask: Optional[torch.Tensor] = None,
        labels: Optional[torch.Tensor] = None,
    ):
        """
        input_ids: input sentences converted to ids
        attention_mask: the attention mask
        labels:  Labels for computing the multiple choice classification loss. Indices should be in `[0, ...,num_choices-1]` where `num_choices` is the size of the second dimension of the input tensors.
        """

        num_choices = input_ids.shape[1]

        ###your code ###

        # flatten input to be able to go through model
        input_ids = input_ids.view(-1, input_ids.size(-1))
        attention_mask = attention_mask.view(-1, attention_mask.size(-1))

        # get last_hidden_state
        dbert_output = self.distilbert(input_ids = input_ids, attention_mask = attention_mask)
        last_hidden_state = dbert_output.last_hidden_state

        # CLS
        cls_emb = last_hidden_state[:, 0, :]

        # layers
        # first linear + activation + dropout
        data = self.dropout(self.activation(self.dense(cls_emb)))

        # second
        data = self.dropout(self.dense2(data))

        # classifier and reshape
        reshaped_logits = self.classifier(data).view(-1, num_choices)

        loss = None
        if labels is not None:
          loss_actual = nn.CrossEntropyLoss()
          loss = loss_actual(reshaped_logits, labels.view(-1))

        ###your code ###
        return {"loss":loss,"logits":reshaped_logits}


Initialize the feature extractor with `distilbert-base-uncased` and create your custome model.

In [ ]:
from transformers import AutoConfig
###your code ###
config = AutoConfig.from_pretrained("distilbert-base-uncased")
model_custom = CustomMultipleChoice(config)
###your code ###

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:72: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


In [ ]:
for name, param in model_custom.named_parameters():
    if param.requires_grad and not name.startswith("distilbert."):
        print(name, param.data.shape)

dense.weight torch.Size([768, 768])
dense.bias torch.Size([768])
dense2.weight torch.Size([192, 768])
dense2.bias torch.Size([192])
classifier.weight torch.Size([1, 192])
classifier.bias torch.Size([1])


We keep the same training arguments but change the directory in which we save the model logs, the directory in which we save the model output and the name of the run, to `custom_model`.



In [ ]:
### your code ###
training_args_custom = TrainingArguments(
    output_dir="./custom_model",
    evaluation_strategy="steps",
    eval_steps=300,
    save_strategy="steps",
    save_steps=600,
    save_total_limit=2,
    seed=77,
    per_device_train_batch_size=48,
    per_device_eval_batch_size=48,
    max_steps=1800,
    learning_rate=5e-5,
    weight_decay=0.01,
    remove_unused_columns=True,
    load_best_model_at_end=True,
)

### your code ###

Initialize the trainer for training the custom model.The training should take around 30 min on Google Colab T4 GPU.


In [ ]:
### your code ###
trainer_custom = Trainer(
    model = model_custom,
    args = training_args_custom,
    train_dataset = encoded_datasets['train'],
    eval_dataset = encoded_datasets['validation'],
    tokenizer = tokenizer,
    data_collator = MultipleChoiceDataCollator(tokenizer),
    compute_metrics = compute_metrics
)
### your code ###

In [ ]:
trainer_custom.train()# should take around 30 min on Colab T4 GPU

TrainOutput(global_step=1800, training_loss=1.3690565490722657, metrics={'train_runtime': 2492.151, 'train_samples_per_second': 34.669, 'train_steps_per_second': 0.722, 'total_flos': 5697484563348576.0, 'train_loss': 1.3690565490722657, 'epoch': 1.17})

Save the model in `custom_model/final_model`. Note that with the custom model, you need to save it without the help of the trainer. The trainer would save the configuration but since this model is not a registered Hugging Face model only the base model would be saved. Loading the model weights is also effected by this.

In [ ]:
###your code ###
path_custom="./custom_model/final_model"
model_custom.save_pretrained(path_custom)
###your code ###

#### ${\color{red}{Comments\ 2.3}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 4: Evaluation and Model Comparison

Many times you do not perform the final evaluation right after training, but load the checkpoints and evaluate them on the fly. To this end, load the two models from  disk.

In [ ]:
from transformers import AutoModelForMultipleChoice,AutoConfig
### your code ###
model_hf = AutoModelForMultipleChoice.from_pretrained("./distilbert-base-uncased-swag/final_model")
model_custom = CustomMultipleChoice.from_pretrained("./custom_model/final_model")
### your code ###

To evaluate the data we load the validation split using a data loader and our previously defined data collator. Note that although we had a test split we cannot use it, since there are no labels available for this split (you can check the data to confirm this).

In [ ]:
from torch.utils.data import DataLoader
import evaluate

eval_dataloader = DataLoader(encoded_datasets["validation"], batch_size=64, collate_fn=MultipleChoiceDataCollator(tokenizer))

To make things easier, let's use the `evaluate` library from Hugging Face to compute the accuracy metric. Here we load `accuracy` from the `evaluate` library two times, one for the custom model and one for the Hugging Face model. Further, we put the models on eval mode. Complete the code for evaluation using the capabilities of the `evaluate` library to simultaneously compute the metric for both models.


In [43]:
from tqdm import tqdm
device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
metric_dict={"custom":evaluate.load("accuracy"),"hf":evaluate.load("accuracy")} #use to compute accuracy
models_dict= {"custom": model_custom, "hf":model_hf}# use to access models

for name, model in models_dict.items():
  model.to(device)
  model.eval()

for i,batch in tqdm(enumerate(eval_dataloader), total=len(eval_dataloader)):
  ### your code ###
  #evaluate on both model on each batch
  input_ids = batch["input_ids"].to(device)
  attention_mask = batch["attention_mask"].to(device)
  labels = batch["labels"].to(device)

  #evaluate first model - Hugging Face
  with torch.no_grad():
    output_hf = model_hf(input_ids, attention_mask = attention_mask, labels = labels)
    logits_hf = output_hf.logits
    predictions_hf = torch.argmax(logits_hf, dim=1)

  #evaluate second model - custom
  with torch.no_grad():
    output_custom = model_custom(input_ids, attention_mask = attention_mask, labels = labels)
    logits_custom = output_custom["logits"]
  predictions_custom = torch.argmax(logits_custom, dim=1)

  #update metrics
  metric_dict["custom"].add_batch(predictions=predictions_custom, references = labels)
  metric_dict["hf"].add_batch(predictions=predictions_hf, references = labels)

acc_hf = metric_dict["hf"].compute()
acc_custom = metric_dict["custom"].compute()
  ### your code ###
print("Hugging Face Model :",acc_hf)
print("Custom Model :",acc_custom)

100%|██████████| 313/313 [04:41<00:00,  1.11it/s]

Hugging Face Model : {'accuracy': 0.7140857742677197}
Custom Model : {'accuracy': 0.2717184844546636}


#### ${\color{red}{Comments\ 2.4}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

## **Task 3: Encoder-Decoder Architecture** (5 + 2 + 2 + 5 = 14 points)

We explored an encoder-based model (BERT) in the previous exercise. In this task, we look at another family of transformer architectures, the encoder-decoder. We use the [T5](https://arxiv.org/pdf/1910.10683.pdf) model, presented by Raffel et al.  T5 is an encoder-decoder architecture pre-trained on a multi-task mixture of unsupervised and supervised tasks. In this task, we set up a fine-tuning example for question answering using the [SQUAD](https://huggingface.co/datasets/squad) dataset. Since the actual fine-tuning is time-consuming and computational intensive for inference, we use an already pre-trained model. The main goal is to introduce you to the structure of the fine-tuning and its simplicity with the Hugging Face framework.

To fine-tune the BERT-based models, we usually add a task-specific head. On the other hand, T5 converts all NLP problems into a text-to-text format.  
It is trained using teacher forcing, meaning that we require an input sequence and a corresponding target sequence.


1.   The input sequence is fed to the model using `input_ids` from the tokenizer.
2.   The target sequence is shifted to the right, i.e., prepended by a start-sequence token and fed to the decoder using the `decoder_input_ids` (input_ids of the encoded target sequence). The target sequence is appended by EOS (end of the sentence) to denote the end of a generation and corresponds to the `labels`.
3. The task prefix defines what task is expected of T5. For example, we prepend the input sequence with `translate English to German: ` before encoding the input to tell the model to translate. T5 already has a set of pre-defined task prefixes, and it is best to stick to those since they were used during pre-training. With enough training data, you can also introduce your own custom task.


In contrast to the encoder model, where only a single `max_length` is required, for encoder-decoder architectures, one typically defines a `max_source_length` and `max_target_length`, which determine the maximum length of the input and output sequences, respectively. We must also ensure that the padding ID of the `labels` is not taken into account by the loss function. This can be done by replacing them with `-100`, which is the `ignore_index` of the `CrossEntropyLoss`.

### Subtask 1: Data Processing

We first start by loading the dataset from Hugging Face hub:

In [ ]:
from datasets import load_dataset

datasets_squad = load_dataset("squad")
datasets_squad

DatasetDict({
    train: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers'],
        num_rows: 87599
    })
    validation: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers'],
        num_rows: 10570
    })
})

In [ ]:
print("context ---->" ,datasets_squad["train"][0]["context"])
print("question ---->",datasets_squad["train"][0]["question"])
print("answers ---->",datasets_squad["train"][0]["answers"])

context ----> Architecturally, the school has a Catholic character. Atop the Main Building's gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary.
question ----> To whom did the Virgin Mary allegedly appear in 1858 in Lourdes France?
answers ----> {'text': ['Saint Bernadette Soubirous'], 'answer_start': [515]}


Now let's load the needed pre-trained tokenizer for `t5-small`, which is the smallest T5 model. Set the maximum sequence length to `512`.

In [ ]:
import torch
### your code ###
from transformers import T5TokenizerFast
t5_tokenizer = T5TokenizerFast.from_pretrained("t5-small", model_max_length=512)
### your code ###

The next step is to pre-process the dataset using the tokenizer to convert the sequences to IDs and add the special tokens.
T5 is based on the SentencePiece tokenizer, and the end of sentence token is denoted by `</s>`.
Complete the function `add_eos_to_examples` to format the input and target sequence. Your input as `input_text` should have the format `question:{question_text} context:{context_text} <EOS_Token>` and your target as `target_text` should have the format `{answer_text} <EOS_Token>`.

In [ ]:
def add_eos_to_examples(example):
    ### your code ###
    example['input_text'] = f"question: {example['question']}  context: {example['context']} </s>"
    example['target_text'] = f"{example['answers']['text'][0]} </s>"
    ### your code ###
    return example

Use the `map` function to process the data, and do not set the `batched` argument.

In [ ]:
### your code ###
encoded_squad = datasets_squad.map(add_eos_to_examples)
### your code ###

Map:   0%|          | 0/87599 [00:00<?, ? examples/s]

Map:   0%|          | 0/10570 [00:00<?, ? examples/s]

In [ ]:
print(encoded_squad["train"][0]["input_text"])
print(encoded_squad["train"][0]["target_text"])

question: To whom did the Virgin Mary allegedly appear in 1858 in Lourdes France?  context: Architecturally, the school has a Catholic character. Atop the Main Building's gold dome is a golden statue of the Virgin Mary. Immediately in front of the Main Building and facing it, is a copper statue of Christ with arms upraised with the legend "Venite Ad Me Omnes". Next to the Main Building is the Basilica of the Sacred Heart. Immediately behind the basilica is the Grotto, a Marian place of prayer and reflection. It is a replica of the grotto at Lourdes, France where the Virgin Mary reputedly appeared to Saint Bernadette Soubirous in 1858. At the end of the main drive (and in a direct line that connects through 3 statues and the Gold Dome), is a simple, modern stone statue of Mary. </s>
Saint Bernadette Soubirous </s>


Complete the function `convert_to_features` that takes in the examples from the dataset and tokenizes them using the T5 tokenizer. However, our answers in this dataset are relatively short and do not require `512` tokens, in contrast to the input sequence which is a combination of question and context paragraphs and is usually long. To this end, we want to truncate the input sequence at `512` and the target sequence at `16`. If any input or target is smaller than the specified length, make sure you pad them. Finally, convert everything to PyTorch tensors to be easily used by the data collator and place them in the dictionary `encodings`.

In [ ]:
def convert_to_features(examples):
    ### your code ###
    input_encodings = t5_tokenizer(examples["input_text"], padding="max_length", truncation=True, max_length=512)
    target_encodings = t5_tokenizer(examples["target_text"], padding="max_length", truncation=True, max_length=16)

    encodings = {
        'input_ids': input_encodings['input_ids'],
        'attention_mask': input_encodings['attention_mask'],
        'target_ids': target_encodings['input_ids'],
        'target_attention_mask': target_encodings['attention_mask']
    }
    ### your code ###
    return encodings

Use the `map` function to process the data.

In [ ]:
### your code ###
encoded_squad = encoded_squad.map(convert_to_features, batched=True)
### your code ###

Map:   0%|          | 0/87599 [00:00<?, ? examples/s]

Map:   0%|          | 0/10570 [00:00<?, ? examples/s]

In [ ]:
encoded_squad #new columns are added

DatasetDict({
    train: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers', 'input_text', 'target_text', 'input_ids', 'attention_mask', 'target_ids', 'target_attention_mask'],
        num_rows: 87599
    })
    validation: Dataset({
        features: ['id', 'title', 'context', 'question', 'answers', 'input_text', 'target_text', 'input_ids', 'attention_mask', 'target_ids', 'target_attention_mask'],
        num_rows: 10570
    })
})

Interestingly, although we specified PyTorch tensors as output, the type of the `input_ids` is still a list. To remedy this problem, you need to explicitly set the type of the column that contains PyTorch tensors.

In [ ]:
type(encoded_squad["train"][0]["input_ids"])

list

In [ ]:
### your code ###
columns = ['input_ids', 'target_ids', 'attention_mask', 'target_attention_mask']
encoded_squad.set_format(type='torch', columns=columns)
### your code ###
type(encoded_squad["train"][0]["input_ids"])

torch.Tensor

In [ ]:
print("Shape of the input_ids:",encoded_squad["train"][0]["input_ids"].shape)
print("Shape of the target_ids:",encoded_squad["train"][0]["target_ids"].shape)

Shape of the input_ids: torch.Size([512])
Shape of the target_ids: torch.Size([16])


The final step in the data processing is the creation of the data collator to
prepare `labels` from `target_ids` and return examples with keys as expected by the forward method of T5.
This is necessary because the trainer directly passes this dict as argument to the model so you need to check the input of T5 and rename the column based on that.
`input_ids`, `target_ids`, `attention_mask`, and `target_attention_mask` need to be stacked in a batch and the pad tokens in the target need to be set to `-100` to avoid loss computation.

In [ ]:
# The final step in the data processing is the creation of the data collator to
# prepare `labels` from `target_ids` and return examples with keys as expected by the forward method of T5.
# This is necessary because the trainer directly passes this dict as argument to the model so you need to check the input of T5 and rename the column based on that.
# `input_ids`, `target_ids`, `attention_mask`, and `target_attention_mask` need to be stacked in a batch and the pad tokens in the target need to be set to `-100` to avoid loss computation.

from dataclasses import dataclass
from transformers import DataCollator
from transformers.tokenization_utils_base import PreTrainedTokenizerBase, PaddingStrategy
from typing import Optional, Union
@dataclass
class T2TDataCollator:
    tokenizer: PreTrainedTokenizerBase
    padding: Union[bool, str, PaddingStrategy] = True
    max_length: Optional[int] = None
    pad_to_multiple_of: Optional[int] = None
    def __call__(self, batch):

      ### your code ###
        input_ids = torch.stack([example['input_ids'] for example in batch])
        attention_mask = torch.stack([example['attention_mask'] for example in batch])

        target_ids = torch.stack([example['target_ids'] for example in batch])
        target_ids[target_ids[:, :] == self.tokenizer.pad_token_id] = -100
        target_attention_mask = torch.stack([example['target_attention_mask'] for example in batch])

        feature_dict = {
            'input_ids': input_ids,
            'attention_mask': attention_mask,
            'labels': target_ids,
            'decoder_attention_mask': target_attention_mask
        }

        return feature_dict
      ### your code ###


In [ ]:
accepted_keys = ['input_text', 'target_text', 'input_ids', 'attention_mask', 'target_ids', 'target_attention_mask']
features = [{k: v for k, v in encoded_squad["train"][i].items() if k in accepted_keys} for i in range(2)]
batch=T2TDataCollator(t5_tokenizer)(features)
print(batch["input_ids"].shape)
print(batch["attention_mask"].shape)
print(batch["labels"].shape)

torch.Size([2, 512])
torch.Size([2, 512])
torch.Size([2, 16])


#### ${\color{red}{Comments\ 3.1}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 2: Training

For training and inference, we can use `T5ForConditionalGeneration`, which includes the language modeling head on top of the decoder. Load the `t5-small` model.

In [ ]:
### your code ###
from transformers import T5ForConditionalGeneration
t5 = T5ForConditionalGeneration.from_pretrained("t5-small")
### your code ###

Next, similar to the previous task we initiate training arguments. Note that this time we are using a `Seq2SeqTrainingArguments` for a `Seq2SeqTrainer`. Set the parameters for training as follows:


*   T5 doesn't support GPU and TPU evaluation for now, so we only focus on training. You do not need to pass any parameters for evaluation setup.
*   The output directory should be named `t5-squad`.
* The T5 models need a slightly higher learning rate than the default one set in the `Trainer` when using the `AdamW` optimizer. Set the learning rate to `1e-4` and the regularization parameter to `0.01`.
* Random seed should be `77`, and we train for a maximum of `200` steps and save a checkpoint every `100` steps. A complete training of the T5 model requires far more than `200` steps, however, that is beyond the scope of this assignment.
* T5 models require a large batch size. The default model was trained with a batch size of `128`. However, we cannot fit that into a single GPU, therefore we use gradient accumulation. Set the batch size to `32` and choose the gradient accumulation step to reach the effective batch size of `128`.
* Make sure that your trainer does not remove unused columns during training, as this will cause a runtime error later on.


**Gradient accumulation:** is a technique that simulates a larger batch size by accumulating gradients from multiple small batches before performing a weight update.



In [ ]:
from transformers import Seq2SeqTrainingArguments

### your code ###
training_args = Seq2SeqTrainingArguments(
    do_train=True,                  # do training
    output_dir="./t5-squad",        # output directory
    learning_rate=1e-4,             # learning rate
    weight_decay=0.01,              # regularization parameter
    seed=77,                        # seed
    data_seed=77,                   # data seed
    max_steps=200,                  # max training steps
    save_steps=100,                 # save every 100 steps
    per_device_train_batch_size=32, # batch size
    gradient_accumulation_steps=4,  # gradient accumulation steps
    remove_unused_columns=False     # do not remove unused columns
)
    ### your code ###


Once again make sure that you are using GPU before running the cell below.
Initilize your `Seq2SeqTrainer` with inputs necessary for training. The training should take around 15 min on Google Colab T4 GPU.


In [ ]:
# Initialize our Trainer
from transformers import Seq2SeqTrainer
### your code ###
trainer = Seq2SeqTrainer(
    model=t5,
    args=training_args,
    data_collator=T2TDataCollator(t5_tokenizer),
    train_dataset=encoded_squad["train"]            # only training, no eval
)
    ### your code ###


In [ ]:
trainer.train()

TrainOutput(global_step=200, training_loss=0.40351577758789064, metrics={'train_runtime': 789.1399, 'train_samples_per_second': 32.44, 'train_steps_per_second': 0.253, 'total_flos': 3464750117683200.0, 'train_loss': 0.40351577758789064, 'epoch': 0.29})

#### ${\color{red}{Comments\ 3.2}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 3: Inference

Our trained model has seen far too few instances to make a coherent prediction. To this end, we load an already trained checkpoint from Hugging Face and perform inference. Load this [model](https://huggingface.co/mrm8488/t5-base-finetuned-squadv2) and the respective tokenizer. Note that we are loading a `base` model that is slightly larger than `t5-small`.

In [ ]:
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
### your code ###
t5_tokenizer = AutoTokenizer.from_pretrained("mrm8488/t5-base-finetuned-squadv2")
t5_model = AutoModelForSeq2SeqLM.from_pretrained("mrm8488/t5-base-finetuned-squadv2")
### your code ###

The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.
The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.
The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.
The `xla_device` argument has been deprecated in v4.4.0 of Transformers. It is ignored and you can safely remove it from your `config.json` file.


At inference time for T5, it is recommended to use the `generate()` function. This auto-regressively generates the decoder output. Complete the code for the `get_answer` function, which gives a model, a tokenizer, and a question and context pair, and generates the answer from the context given. The output should be the answer to the given question in natural text (without the special tokens).

**Hint:** Many of the steps are similar to how you prepared your input data for the model.

In [ ]:
def get_answer(tokenizer,model, question, context):
    ### your code ###
    input_text = f"question: {question}  context: {context} </s>"
    features = tokenizer([input_text], return_tensors='pt', max_length=512, truncation=True, padding="max_length")

    answers_encoded = model.generate(input_ids=features['input_ids'],
                                     attention_mask=features['attention_mask'],
                                     max_length=16)
    answer = tokenizer.decode(answers_encoded[0], skip_special_tokens=True)
    ### your code ###
    return answer

Let's try it with an example.

In [ ]:
context = "Sarah has joined NLP for transformers class and is working on her research project with the support of Harry."
question = "Who is supporting Sarah?"

get_answer(t5_tokenizer,t5_model,question, context)###your answer should be "Harry"

'Harry'

In [ ]:
context = "TPUs are more power efficient in comparison to GPUs making them a better choice for machine learning projects."
question = "What is better for machine learning projects?"

get_answer(t5_tokenizer,t5_model,question, context)###your answer should be "TPUs"

'power efficient'

#### ${\color{red}{Comments\ 3.3}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$

### Subtask 4: T5 Paper

To answer questions of the final subtask you need to have a general overview of the [T5 paper](https://arxiv.org/pdf/1910.10683.pdf).



1.   Describe what a “text-to-text format" is and how T5 processes input and output for text classification tasks? What are the possible complications with a predefined set of classes?
2.   Describe the "masked language modeling" and "word dropout" unsupervised objective with sentinel tokens. Give an example of how this would look in a single sentence.
3. Explain "fully-visible", "causal" and "causal masking with prefix" masking.
4. Briefly describe "adapter layers" and "gradual unfreezing" as methods for fine-tuning on fewer parameters.



**Answer**

```
1.1 - What is a "Text-to-text format"?

"Text-to-text format" is about treating "every text processing problem as (...) taking text as input and producing new text as output". In other words, the authors came up with an idea to reformulate different text-related tasks (question answering, summarization and others) into text-to-text task which allowed them to unify the model, training, decoding etc.

1.2 - How T5 processes input and output?

T5 has a quite symmetrical architecture consisting of an encoder and decoder. To tell the model what the current task is, each input begins with predefined prefix such as "summarize: " or "translate English to German: ". This prefix is in natural language and is not treated separately from the rest of the input. Number of these prefixes was fixed during training. For text classification in particular, the model has to predict "a single word corresponding to the target label".

1.3 - What are the possible complications with a predefined set of classes?

Model might be unable to generalize to arbitrary tasks which were not present during trainig. Additionally, the paper says that only specific languages are supported by the vocabulary which suggests that changing the language of the objective might also negatively impact model behavior.
```

**Answer**

```
2.1 - Describe the "masked language modeling" and "word dropout" unsupervised objective with sentinel tokens.

The authors desribe it as inspired by BERT and Bowman et al. (2015). In this objective, in general some tokens are removed and the target is to recreate the original input.

In more details, the procedure is described as follows:
- randomly sample and drop 15% of tokens in the input sequence
- replace all consecutive spans of dropped-out tokens with a single sentinel token
- assign a unique token ID to each sentinel token (sentinel token IDs are special - they are added to the vocabulary but do not correspond to any wordpiece)
- the target is to recreate the original input with all dropped-out spans of tokens delimited by the same sentinel tokens used in the input sequence plus a final sentinel token to mark the end of the target sequence.

Additionally, in a later part, the authors mention that in their training procedure masked language modelling "takes a span of text and corrupts 15% of the tokens. 90% of the corrupted tokens are replaced with a special mask token and 10% are replaced with a random token." This would suggest that not all masked tokens are replaced with sentinel tokens, if my understanding is correct.


2.2 - Give an example of how this would look in a single sentence.

My sample sentence has 20 tokens (here, I treat words as tokens), so I should only mask 15% of them which is 3. To make the example more meaningful, I will mask more tokens and replace them with sentinel tokens. I will not replace 10% of masked tokens with a random token as the authors did during training. Based on my understanding presented above and taking the following sentence as an example (for readability, each word here will be treated as a single token).

Original:

"My name is John and I am a student at the old University of Heidelberg which is located in Germany"

Modified:

"My name is [SENTINEL ID 1] student at the old [SENTINEL ID 2] of Heidelberg which is located in Germany [FINAL SENTINEL TOKEN]"

The target would be to recreate, that:
- [SENTINEL ID 1] stands for: "John and I am a" (here, a whole span of consecutive tokens is replaced with a single sentinel token)
- [SENTINEL ID 2] stands for: "University" (but a single dropped-out word is treated the same)
- [FINAL SENTINEL TOKEN] marks the end of the target sequence
```

**Answer**

```
3 - Explain "fully-visible", "causal" and "causal masking with prefix" masking.

These are different attention mask patterns, as the Figure 3 nicely visualized.

- Fully-visible masking scheme does not mask anything and "allows the self-attention mechanism to attend to the full input at every output timestep".

- A causal mask is such that it prevents the self-attention mechanism from attending to any tokens that occur after the current output timestep (prevents 'looking into the future').

- Causal masking with prefix is a combination of the two above - first few tokens from the input (a prefix) are visible all the time (even for the first few output timesteps) and the rest of the input is masked (as in causal masking). A good example would be with “translate English to German: That is good. target: Das ist gut.", where the last "Das ist gut." would be masked with casual masking and the first part would be a fully visible prefix.
```

**Answer**

```
4 - Briefly describe "adapter layers" and "gradual unfreezing" as methods for fine-tuning on fewer parameters.

- "adapter layers" approach is about inserting additional small neural networks (dense-ReLU-dense blocks with size d as hyperparameter) with same input and output sizes "after each of the preexisting feed-forward networks in each block of the Transformer." During finetuning, only these small networks are trained, while the rest of the model is frozen.

- "gradual unfreezing" is about starting finetuning of the model by finetuning only the last layer and, later, gradually unfreezing additional layers (going from the last layer to the first one). The authors used an encoder-decoder model and they were unfreezing layers in the encoder and decoder in parallel (from the top in both cases).
```

#### ${\color{red}{Comments\ 3.4}}$

${\color{red}{⚠️Comments\ begin⚠️}}$


```
cross-feedback comment section
```


${\color{red}{⚠️Comments\ end⚠️}}$